pages/event.php

```diff
@@ -189,16 +228,228 @@ function event() {
+                case "shadow_sacred_violet":
+                    $player->getInventory();
+                    if (!$player->hasItem($system->event_data['shadow_essence_id'])) {
+                        throw new RuntimeException("You do not have this item!");
+                    }
+                    if ($player->items[$system->event_data['shadow_essence_id']]->quantity < 1) {
+                        throw new RuntimeException("You do not have enough of this item!");
+                    }
+                    if ($player->hasItem($system->event_data['sacred_lantern_violet'])) {
+                        throw new RuntimeException("You already have this item!");
+                    }
+                    $player->items[$system->event_data['shadow_essence_id']]->quantity -= 1;
+                    if ($player->items[$system->event_data['shadow_essence_id']]->quantity < 1) {
+                        unset($player->items[$system->event_data['shadow_essence_id']]);
+                    }
+
+
+                    $result = $system->db->query("SELECT * FROM `items` WHERE `item_id` = {$system->event_data['sacred_lantern_violet']}");
+                    $player->items[$system->event_data['sacred_lantern_violet']] = Item::fromDb($system->db->fetch($result), 1);
+
+                    $system->message("You have traded 1 Shadow Essence for a Sacred Violet Lantern!");
+                    $player->updateInventory();
+                    $player->updateData();
+                    break;
+                case "shadow_jutsu":
+                    $player->getInventory();
+                    if (!$player->hasItem($system->event_data['shadow_essence_id'])) {
+                        throw new RuntimeException("You do not have this item!");
+                    }
+                    if ($player->items[$system->event_data['shadow_essence_id']]->quantity < 1) {
+                        throw new RuntimeException("You do not have enough of this item!");
+                    }
+                    if ($player->hasItem($system->event_data['forbidden_jutsu_scroll'])) {
+                        throw new RuntimeException("You already have this item!");
+                    }
+                    $player->items[$system->event_data['shadow_essence_id']]->quantity -= 1;
+                    if ($player->items[$system->event_data['shadow_essence_id']]->quantity < 1) {
+                        unset($player->items[$system->event_data['shadow_essence_id']]);
+                    }
+
+                    $result = $system->db->query("SELECT * FROM `items` WHERE `item_id` = {$system->event_data['forbidden_jutsu_scroll']}");
+                    $player->items[$system->event_data['forbidden_jutsu_scroll']] = Item::fromDb($system->db->fetch($result), 1);
+
+                    $system->message("You have traded 1 Shadow Essence for a Forbidden Jutsu Scroll!");
+                    $player->updateInventory();
+                    $player->updateData();
```

Comment:
There seems to be a lot of opportunity to extract common logic here into shared method(s) but I don't want to block this, we can do that later